primitives/src/radio_group.rs

```diff
@@ -0,0 +1,228 @@
+        if let Some(current_focus) = (self.current_focus)() {
+            let mut new_focus = current_focus.saturating_add(1);
+
+            let item_count = (self.item_count)();
+            if new_focus >= item_count {
+                match (self.roving_loop)() {
+                    true => new_focus = 0,
+                    false => new_focus = item_count.saturating_sub(1),
+                }
+            }
+
+            self.current_focus.set(Some(new_focus));
+        }
+    }
+
+    fn focus_prev(&mut self) {
+        if let Some(current_focus) = (self.current_focus)() {
+            let mut new_focus = current_focus.saturating_sub(1);
+            if current_focus == 0 && (self.roving_loop)() {
+                new_focus = (self.item_count)().saturating_sub(1);
+            }
+
+            self.current_focus.set(Some(new_focus));
+        }
+    }
+
+    fn focus_start(&mut self) {
+        self.current_focus.set(Some(0));
+    }
+
+    fn focus_end(&mut self) {
+        let new_focus = (self.item_count)().saturating_sub(1);
+        self.current_focus.set(Some(new_focus));
+    }
+}
+
+#[derive(Props, Clone, PartialEq)]
+pub struct RadioGroupProps {
+    value: Option<Signal<String>>,
+
+    #[props(default)]
+    default_value: String,
+
+    #[props(default)]
+    on_value_change: Callback<String>,
+
+    #[props(default)]
+    disabled: ReadOnlySignal<bool>,
+
+    #[props(default)]
+    required: ReadOnlySignal<bool>,
+
+    #[props(default)]
+    name: ReadOnlySignal<String>,
+
+    #[props(default)]
+    horizontal: ReadOnlySignal<bool>,
+
+    #[props(default = ReadOnlySignal::new(Signal::new(true)))]
+    roving_focus: ReadOnlySignal<bool>,
+
+    #[props(default = ReadOnlySignal::new(Signal::new(true)))]
+    roving_loop: ReadOnlySignal<bool>,
+
+    #[props(extends = GlobalAttributes)]
+    attributes: Vec<Attribute>,
+
+    children: Element,
+}
+
+#[component]
+pub fn RadioGroup(props: RadioGroupProps) -> Element {
+    let (value, set_value) =
+        use_controlled(props.value, props.default_value, props.on_value_change);
+
+    let mut ctx = use_context_provider(|| RadioGroupCtx {
+        value: value.into(),
+        set_value,
+        disabled: props.disabled,
+
+        item_count: Signal::new(0),
+        recent_focus: Signal::new(0),
+        current_focus: Signal::new(None),
+        horizontal: props.horizontal,
+        roving_focus: props.roving_focus,
+        roving_loop: props.roving_loop,
+    });
+
+    rsx! {
+        div {
+            role: "radiogroup",
+            "data-orientation": if (props.horizontal)() { "horizontal" } else { "vertical" },
+            "data-disabled": (props.disabled)(),
+            aria_required: props.required,
+
+            onfocusout: move |_| ctx.set_focus(None),
+            ..props.attributes,
+
+            {props.children}
+        }
+    }
+}
+
+#[derive(Props, Clone, PartialEq)]
+pub struct RadioItemProps {
+    value: String,
+    index: usize,
```

Comment:
Also here should be a `ReadOnlySignal`. While we may not need it to be a signal, I'd prefer it as it gaurantees reactivity.